### PR TITLE
Allow mocha callbacks to use any subclass of Context

### PR DESCRIPTION
**Pull request by joeyparrish**
When creating fixtures in Mocha, they get attached to Context, which includes an index signature for this purpose.  However, those fixtures will have no type information.

If a project wants to strongly type its fixtures, it needs to subclass Context.  But then, without this patch, the compiler will reject the use of those subclasses in Mocha callbacks.

---

**Update**: Updated again with latest design changes to the PR.

`Context`'s index signature prevents strong typing for Mocha fixtures.  To address this, a project should subclass `Context` and enable the compiler option `noPropertyAccessFromIndexSignature`, but the mocha types still need to use generics to allow the subclass for the type of `this` in callbacks.

Callback functions `Func` and `AsyncFunc` had `this: Context`, but now are generics with `this: T` and `T extends Context = Context`.

Leaving out the default `T = Context` breaks compilation, with errors like `Generic type 'AsyncFunc' requires 1 type argument(s)`.  But including it also leads to the assumption that we always pass `Context` instead of a subclass.

The solution is for each usage of `Func` and `AsyncFunc` in Mocha's API, to add a generic overload of the method.  If I pass a callback that uses strongly-typed fixtures in a `Context` subclass, the compiler can infer the generic type and properly check my usage of those fixtures.

For example, where we had this:

```ts
            before(fn?: Mocha.Func | Mocha.AsyncFunc): void;
```

we change to this:

```ts
            before<T extends Mocha.Context>(fn?: Mocha.Func<T> | Mocha.AsyncFunc<T>): void;
```

Each instance of this fails the linter check `@definitelytyped/no-unnecessary-generics` with `Type parameter T is used only once`.  The generic doesn't influence the return type, so this rule would steer us toward using `any` instead.  But this is clearly necessary to use the strongly-typed `this` we need to pass.  So each instance is preceded by `// eslint-disable-next-line @definitelytyped/no-unnecessary-generics`.

---

Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://mochajs.org/#global-fixtures
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

**Comment by joeyparrish** [comment]
Everything is passing and I'm looking forward to your feedback!

**Comment by joeyparrish** [comment]
I know this is 100+ lines changed, but there are really only 2-3 things happening, one of them repeated many, many times.  To aid review, I updated the PR description with a better explanation of the change.  It details the things that changed, why these are necessary changes, and why the simpler changes I tried don't work.  I hope this helps!

File: types/mocha/index.d.ts
Comment (rbuckton) [inline types/mocha/index.d.ts:393]: Would `<T extends BaseContext = Context>(fn: Func<T>): void;` allow you to remove the non-generic version?

Comment (joeyparrish) [reply types/mocha/index.d.ts:393]: No, because then every time I invoke the function, I would have to call it with an explicit subtype, or the compiler would assume Context.  This way, the compiler infers the type.  It's ugly, I know.

File: types/mocha/index.d.ts
Comment (jakebailey) [inline types/mocha/index.d.ts:1342]: Does this class actually exist at runtime? This will be exported so implies it can be new'd.

Also, the comment above is attached to the wrong declaration.

**Comment by sandersn** [comment]
1. I'd love to have a review from one of the original owners: @kazimanzurrashid, @otiai10, @enlight, @cspotcode, @1999, @strangedev, @nicojs, although it's possible that they have all moved on.
2. mocha is so widely used that I suspect any changes will cause breaks for *somebody*. Is there a way to attach this change to a major version bump of mocha itself?

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
Waiting on reviews from original owners per https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69255#issuecomment-2115316142

File: types/mocha/index.d.ts
Comment (joeyparrish) [reply types/mocha/index.d.ts:1342]: No, it does not exist.  But this doesn't work with BaseContext as an interface.

To have a strongly-typed extension of a Mocha Context, a base class definition must exist without the index signature, so that types in the extension can have strong type enforcement.  Extending `Context` with its index signature means a field-name typo on the subclass is ignored.

Making BaseContext an interface (no implied constructor) doesn't work at all.  The generic calls that reference it say `T extends BaseContext`, but `T implements BaseContext` doesn't work.

I've fixed the placement of the comment you mentioned, and expanded comments on both BaseContext and Context to clarify their purpose.

I've also added a protected constructor to the definition of BaseContext so that it can't be constructed directly, and I added a new test to verify that.

**Comment by joeyparrish** [comment]
> 1. I'd love to have a review from one of the original owners: @kazimanzurrashid, @otiai10, @enlight, @cspotcode, @1999, @strangedev, @nicojs, although it's possible that they have all moved on.

I would love that, too!

> 2. mocha is so widely used that I suspect any changes will cause breaks for _somebody_.

That's what tests are for!  I have made mistakes in this process that caused tests to fail in related projects, such as mochawesome.  Those have been fixed, so the network of tests on related projects appears to be WAI.

> Is there a way to attach this change to a major version bump of mocha itself?

No, because we're not changing Mocha.  There's no new release of Mocha here.  We're just splitting out the index-signature so that extensions can have strong types.

File: types/mocha/index.d.ts
Comment (rbuckton) [reply types/mocha/index.d.ts:1349]: `extends` in a type parameter is not the same as `extends` in a class declaration. In a type parameter it defines a constraint and doesn't care if that constraint is a class or interface.

Comment (jakebailey) [reply types/mocha/index.d.ts:1349]: Yeah, I don't know what you mean here. You can just write:

```ts
interface BaseContext {} 

class Context implements BaseContext {}

beforeAll<T extends BaseContext>(fn?: Func<T>): this;
```

**Comment by jakebailey** [comment]
I still feel moderately confused as how this PR is needed; do you have some concrete motivating code I could look at? The lint rule is trying to steer the types away from potential unsafety, which this PR definitely would introduce.

File: types/mocha/index.d.ts
Comment (joeyparrish) [reply types/mocha/index.d.ts:1349]: @jakebailey, I actually tried that (interface BaseContext, class Context implements BaseContext).  But if we do that, then we have to duplicate every member of that interface in Context, or else we get:

```
compile error: 
Class 'Context' incorrectly implements interface 'BaseContext'.
Type 'Context' is missing the following properties from type 'BaseContext': ...
```

I think it's preferable not to duplicate them, because future methods could be accidentally added to Context without adding them also to BaseContext.  By lying about BaseContext being a class, we can avoid repeating ourselves and avoid letting the definitions get out of sync in the future.

Thoughts?

Comment (jakebailey) [reply types/mocha/index.d.ts:1349]: Where did you test that? IIRC that kind of error should not appear in a declaration file.

image

**Comment by joeyparrish** [comment]
> I still feel moderately confused as how this PR is needed; do you have some concrete motivating code I could look at?

For a concrete example, I added a test that closely represents what I'm doing in my project.  (My project is not yet public.)  The short version is this: I want strongly-typed fixtures, so that typos are caught at compile-time.  The index-signature prevents this.  Here's the test (already in this PR) that shows what I'm after:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/7769552d07da329b899ca465fd4a6042b812565e/types/mocha/mocha-tests.ts#L1481-L1485

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/7769552d07da329b899ca465fd4a6042b812565e/types/mocha/mocha-tests.ts#L1502-L1509

If I pass the wrong arguments, or if I use a name that isn't defined on my class, or if I typo a name, I want the compiler to stop me.  But the index-signature on Context means anything goes.  Possible solutions (by no means exhaustive):

1. Delete the index-signature from Context, potentially breaking existing applications that extend Context without a subclass
2. My subclass tells the compiler to "delete" the index-signature in that subclass (like "deleted functions" [1], [2] in C++; does not exist in TS AFAICT)
3. My subclass inherits from a base that doesn't include the index-signature (the solution I chose)

I am **100%** open to other solutions to this problem, and I am happy to spend my time to implement alternatives that would be more acceptable to the maintainers.  I don't want to be stuck with a fork forever.

Does anyone have other suggestions I could try?

[1]: https://www.learncpp.com/cpp-tutorial/deleting-functions/
[2]: https://www.ibm.com/docs/ru/i/7.4?topic=definitions-deleted-functions-c11

-----

> The lint rule is trying to steer the types away from potential unsafety, which this PR definitely would introduce.

I don't know what you mean by this.  Can you explain?

My view is that this introduces a new way to achieve additional type-safety (through subclassing BaseContext as in my test code).  Without this, I **can't** have type-safe fixtures in Mocha.  How am I introducing potential unsafety?

The change I'm proposing is backwards-compatible for projects that don't require the strict fixture safety that I do.  The evidence is in the test passes.  Previous iterations had mistakes that caused failures in tests for packages that depend on mocha, such as mochawesome.  All packages, not just mocha, are passing with these changes.

File: types/mocha/index.d.ts
Comment (jakebailey) [reply types/mocha/index.d.ts:1349]: Agh, I had `skipLibCheck` on.

If we need to lie about it, then we need to make sure that the base class is unexported. Otherwise, the library is lying about there being a runtime value there when there is not.

But, I'm not sure it's such a bad thing to have to declare things in the interface? Can't a comment be left?

Comment (jakebailey) [reply types/mocha/index.d.ts:1349]: (given mocha's new maintainership, I do also wonder if this will long-term come back to bite us, given maybe it will be converted to TS itself and make these hacks go away, but maybe not)

Comment (joeyparrish) [reply types/mocha/index.d.ts:1349]: > But, I'm not sure it's such a bad thing to have to declare things in the interface? Can't a comment be left?

Sorry, could you be more explicit?  Are you suggesting we duplicate every method in BaseContext and Context and just leave a comment that people should populate/update both?

If that's what you mean, I don't think one comment would suffice.  If I were adding a missing method without knowing the structure, I would go straight to the bottom of Context and add it.  To know about this, I would probably need to see a comment on _every_ method mentioning its duplicate.  That feels heavy to me.

Ultimately, I want to see this merged, and I **am** willing to compromise on these things if y'all feel strongly.

**Comment by jakebailey** [comment]
> I don't know what you mean by this. Can you explain?

Any function that looks like this:

```ts
function foo<T>(): T
```

Is a hazard because it allows you to put _whatever_ you want in as the type argument. People often ask why `JSON.parse` isn't `parse<T>(value: string): T`, and that's because you can do basically whatever you want with it.

You may think that having the sole argument be in the parameter makes it safer:

```ts
interface Something {
    func<T extends BaseContext>(fn: Func<T>): void;
}
```


But, that just means I can write code like:

```
declare const s: Something;

s.func((arg: WhateverIWant) => {})
```

Without having _anything_ that determines that it's legal. It's like casting, but without any syntax that shows that there's actually an unsafety.

File: types/mocha/index.d.ts
Comment (joeyparrish) [reply types/mocha/index.d.ts:1349]: > (given mocha's new maintainership, I do also wonder if this will long-term come back to bite us, given maybe it will be converted to TS itself and make these hacks go away, but maybe not)

If Mocha goes to TS, then DefinitelyTyped won't need to have separate types, right?  If that happens, I can argue for a better structure upstream that accommodates strong types, maybe without an index-signature in the first place.  While Mocha is written in JS, though, the index-signature is implied by the nature of JavaScript.

**Comment by jakebailey** [comment]
But, maybe it's just plain unavoidable.

Notably, ava forces you to do the cast: https://github.com/avajs/ava/blob/f8bf00cd988b5e981b6c7d87523a1e0c5dc947c0/examples/typescript-context/source/test.ts#L5

File: types/mocha/index.d.ts
Comment (jakebailey) [reply types/mocha/index.d.ts:1349]: > If Mocha goes to TS, then DefinitelyTyped won't need to have separate types, right?

Mainly I just meant that while we can "lie" here about there being a class, a real project in TS can't do that.

**Comment by jakebailey** [comment]
I'll take a harder look at the examples you put in the PR if they're complete, though I don't have a lot of time today/tomorrow to do so.

**Comment by joeyparrish** [comment]
> > I don't know what you mean by this. Can you explain?
> 
> Any function that looks like this:
> 
> ```ts
> function foo<T>(): T
> ```
> 
> Is a hazard because it allows you to put _whatever_ you want in as the type argument.

But I didn't do that.  I am always using constraints of the form:

```ts
<T extends BaseContext>(fn: Func<T>): void;
```

> You may think that having the sole argument be in the parameter makes it safer:
> 
> ```ts
> interface Something {
>     func<T extends BaseContext>(fn: Func<T>): void;
> }
> ```
> 
> But, that just means I can write code like:
> 
> ```
> declare const s: Something;
> 
> s.func((arg: WhateverIWant) => {})
> ```
> 
> Without having _anything_ that determines that it's legal. It's like casting, but without any syntax that shows that there's actually an unsafety.

I believe you are mistaken.  We can test this by making your example more concrete.  An example of `Something` in this PR is `Suite`, `func` is `beforeAll`, and `arg` is `this`.  Here's what your example would look like using the actual definitions in this PR:

```ts
// You should not be able to pass arbitrary "this" types.
function test_no_arbitrary_function_arguments(s: LocalMocha.Suite) {
    // Should work.
    s.beforeAll(function(this: LocalMocha.Context) {});

    // Should not work.
    // @ts-expect-error
    s.beforeAll(function(this: string) {});
}
```

And indeed, the second `beforeAll` correctly fails with `No overload matches this call.`

I will add this to the test suite if there are no objections.  Does this help satisfy your concerns?

**Comment by joeyparrish** [comment]
Oh, wait, I was wrong.  It **does** let me get away with it if I do this instead of `string`:

```ts
class ArbitraryClass {}

// ...

    s.beforeAll(function(this: ArbitraryClass) {});
```

Okay, point taken.

**Comment by jakebailey** [comment]
Right; you can pass in whatever you want so long as it fits the constraint. Now, that may be "fine" for a codebase like this where you're intending to lie about the context being totally there and complete; my ava example with the cast "lies" and you have to actually assign to the context somewhere.

Or, it just "has" to be that way because you're declaring a mock lib or something:

```ts
declare function mock<T>(): T;

const mockedFoo = mock<Foo>();
```

> But I didn't do that. I am always using constraints of the form:

Just to be clear, I was using `<T>() => T` as the simplest example; I can just as easily write:

```ts
declare function doSomething<T extends {}>(x: T): void;
```

With a constraint and get feedback when I'm attempting to pass in something bad like `undefined`. It's just that it's a little pointless given I could have just written `doSomething(x: {})`.

In your PRs case, it was a similar question of "why isn't `(fn: Func): void` sufficient".

**Comment by joeyparrish** [comment]
I am flabbergasted.  It seems constraints don't matter at all.  All of these pass, but only f1 makes sense:

```ts
class Base {}
class Sub extends Base {}
class ArbitraryClass {}

type Func<T extends Base> = (this: T) => void;

function test_wtf() {
  // Should work
  Func<Sub> f1 = function(this: Sub) {};

  // Should fail constraints
  Func<ArbitraryClass> f2 = function(this: ArbitraryClass) {};

  // Should fail constraints
  Func<string> f3 = function(this: string) {};

  // Not even self-consistent...
  Func<Sub> f4 = function(this: ArbitraryClass) {};
}
```

What am I doing wrong?

**Comment by jakebailey** [comment]
For completeness, your snippet does work so long as the classes are not structurally identical: [Playground Link](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEBCkFNoG8CwAoa0DEB9AXNGAHYCeA3BgL4aiQwDKArgEbQIAeALgsQCYx4EJGkw4CRMpXQ10dKNACCAJxYBLLsrDLSAYXALRWPIRIVqGDF1IAHJADEmxYAB4AKu268BcRAD5oAF5oAAouAAs1CEI3AEoggIA3AHs1PmkMADMnYC41ZOJoHgguXAB3LkyQ+KNoAHo66AZw5KYQPmgy5OUAawwsYAKS6EyARkJHZxdmFgDg7Oc8grDI6KbWGqoMsQamlraOzLA1EGhB4hKtNWIuCH6zoa4RgCYJnJcVdU1tPQMIOZGOSWxBWUUInw0Wh0+nom22WF2zVa7RGx1O50uxxudzEGKemQAzG8ppdrgBzAELXL5EERMHQUnEMlwyw7RoAOWSTwQiV4DIQIEyAFoMVEeDcAHRS+54kYAFmJrhmlKBNNBawh32hfxZMiAA)

**Comment by joeyparrish** [comment]
> For completeness, your snippet does work so long as the classes are not structurally identical: [Playground Link](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEBCkFNoG8CwAoa0DEB9AXNGAHYCeA3BgL4aiQwDKArgEbQIAeALgsQCYx4EJGkw4CRMpXQ10dKNACCAJxYBLLsrDLSAYXALRWPIRIVqGDF1IAHJADEmxYAB4AKu268BcRAD5oAF5oAAouAAs1CEI3AEoggIA3AHs1PmkMADMnYC41ZOJoHgguXAB3LkyQ+KNoAHo66AZw5KYQPmgy5OUAawwsYAKS6EyARkJHZxdmFgDg7Oc8grDI6KbWGqoMsQamlraOzLA1EGhB4hKtNWIuCH6zoa4RgCYJnJcVdU1tPQMIOZGOSWxBWUUInw0Wh0+nom22WF2zVa7RGx1O50uxxudzEGKemQAzG8ppdrgBzAELXL5EERMHQUnEMlwyw7RoAOWSTwQiV4DIQIEyAFoMVEeDcAHRS+54kYAFmJrhmlKBNNBawh32hfxZMiAA)

Ah, wow, I didn't realize TS was actually comparing structures there.

**Comment by joeyparrish** [comment]
Okay, so after playing with things a bit, it seems that a completely empty class is treated as matching all constraints.  I don't understand why that should be, but it seems like it's a TypeScript issue.  If ArbitraryClass has even one field, it will be checked against constraints and fail with `No overload matches this call` for a call or `missing the following properties from type 'BaseContext'` in an assignment.

Based on that, I **think** this is worth adding to the test suite:

```ts
// Importantly, this doesn't implement BaseContext, and it isn't empty.  For
// some reason, TS treats empty classes as matching constraints.
class ArbitraryClass {
    someField: number;
}

// Context constraints should be enforced on callbacks.
function test_constraints(suite: LocalMocha.Suite) {
    // Should work.
    const f1: LocalMocha.Func<LocalMocha.Context> = function(this: LocalMocha.Context) {};

    // Should work.
    const f2: LocalMocha.Func<ContextWithFixtures> = function(this: ContextWithFixtures) {};

    // Doesn't implement BaseContext.
    // @ts-expect-error
    const f3: LocalMocha.Func<ArbitraryClass> = function(this: ArbitraryClass) {};

    // Doesn't implement BaseContext.
    // @ts-expect-error
    const f4: LocalMocha.Func<string> = function(this: string) {};

    // Doesn't match types.
    // @ts-expect-error
    const f5: LocalMocha.Func<LocalMocha.Context> = function(this: ArbitraryClass) {};

    // Should work.
    suite.beforeAll(function(this: LocalMocha.Context) {});

    // Should work.
    suite.beforeAll(function(this: ContextWithFixtures) {});

    // Doesn't implement BaseContext.
    // @ts-expect-error
    suite.beforeAll(function(this: ArbitraryClass) {});

    // Doesn't implement BaseContext.
    // @ts-expect-error
    suite.beforeAll(function(this: string) {});
}
```

Does this help?

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
Needs rework per feedback

**Comment by jakebailey** [comment]
Er, not yet, I haven't yet looked at the motivating examples...

**Comment by joeyparrish** [comment]
Looks like I can simplify this with [`noPropertyAccessFromIndexSignature`](https://www.typescriptlang.org/tsconfig/#noPropertyAccessFromIndexSignature).  That compiler option eliminates my issue with the index-signature, allowing me to undo the BaseContext/Context split I introduced, which is the truly "fictional" and somewhat controversial part of this PR.

But I still need the introduced generics in order to do this:

```ts
interface TestContext extends Mocha.Context {
  myFixture: Whatever;
}

describe('thing', function() {
  it('does a thing', async function(this: TestContext) {
  });
});
```

Otherwise, I get errors like:

```
error TS2769: No overload matches this call.
  Overload 1 of 4, '(title: string, fn?: Func | undefined): Test', gave the following error.
    Argument of type '(this: TestContext) => Promise<void>' is not assignable to parameter of type 'Func'.
      The 'this' types of each signature are incompatible.
        Type 'Context' is not assignable to type 'TestContext'.
  Overload 2 of 4, '(title: string, fn?: AsyncFunc | undefined): Test', gave the following error.
    Argument of type '(this: TestContext) => Promise<void>' is not assignable to parameter of type 'AsyncFunc'.
      The 'this' types of each signature are incompatible.
        Type 'Context' is not assignable to type 'TestContext'.
```

@jakebailey, how about I undo the `BaseContext` part of the PR and adjust/add to my test cases?

**Comment by jakebailey** [comment]
I guess I don't understand why that `noPropertyAccessFromIndexSignature` helps; that rule is effectively a lint rule and does not really affect the correctness of any code (I don't think it would be added to TS today if it hadn't already existed).

That error is also still pointing out the unsafety I mentioned earlier, FWIW. It's trying to tell you that there's no guarantee that the `this` context will be a narrower type than `Context` itself, and the typical thing to do here is to cast, not allow code that can fail. Same as in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69255#issuecomment-2263663619 / https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69255#issuecomment-2263711227.

**Comment by joeyparrish** [comment]
No, that error is what I get _without_ my PR.  I'll push updates and update the PR description accordingly.  It's much simpler now, and I think you'll find it more acceptable.  These latest changes address your biggest concern as I understood it, and I am still getting everything I need with the simpler PR + `noPropertyAccessFromIndexSignature`.

**Comment by jakebailey** [comment]
> No, that error is what I get _without_ my PR. I'll push updates and update the PR description accordingly.

That is exactly what I'm saying; this is code that _could_ crash because there's no guarantee that any of these contexts are _actually_ being used. I can invent any context I want, the code will typecheck, and nothing will complain about runtime errors. At least when it's `Context`, I'm forced to cast and show TS that what I'm doing is unsafe. That's the core problem with the PR on the whole.

**Comment by joeyparrish** [comment]
> I guess I don't understand why that `noPropertyAccessFromIndexSignature` helps; that rule is effectively a lint rule and does not really affect the correctness of any code (I don't think it would be added to TS today if it hadn't already existed).

The reason I split `Context` into `BaseContext` was to get rid of the index signature, but `noPropertyAccessFromIndexSignature` means I don't need to do that.  I can enable `noPropertyAccessFromIndexSignature`, which catches accidental typos that would otherwise be okay with the index signature.  So no more reason for me to introduce `BaseContext`, which gets rid of the big lie in the PR, and no more reason to keep two versions of every function (standard `Context` and generic `T extends BaseContext`).  So it's much simpler now that I know about `noPropertyAccessFromIndexSignature`, because I don't have to jump mocha types through hoops to achieve the same effect.

Sorry for the all churn and complexity, but I think the new version of this PR is _much_ simpler.  Please take a look!

**Comment by jakebailey** [comment]
But, that just means everyone has to enable that flag to get the benefit? That flag is not very popular, I'm pretty sure.

**Comment by joeyparrish** [comment]
> But, that just means everyone has to enable that flag to get the benefit? That flag is not very popular, I'm pretty sure.

That's fine.  I had two needs.  I was trying to get both from this one PR.  Now I can get one from a simpler, less controversial change, and one from a compiler option.  It doesn't matter to me whether other people have the same needs.  It's all backwards compatible, and this change doesn't hurt anyone.  I don't need to make other people care about this index-signature issue, I just have to have a solution for myself so I can be strict in my project.

-----

> That is exactly what I'm saying; this is code that _could_ crash because there's no guarantee that any of these contexts are _actually_ being used. I can invent any context I want, the code will typecheck, and nothing will complain about runtime errors. At least when it's `Context`, I'm forced to cast and show TS that what I'm doing is unsafe. That's the core problem with the PR on the whole.

So you're saying that this (casting):

```ts
interface MyContext extends Mocha.Context {
  doSomething(): void;
}

it('does something', function() {
  const context = this as MyContext;
  context.doSomething();
});
```

Is better than this (generic, extended types):

```ts
interface MyContext extends Mocha.Context {
  doSomething(): void;
}

it('does something', function(this: MyContext) {
  this.doSomething();
});
```

Do I understand you correctly?

Either way, am I not declaring that I know what I'm doing by stating the type explicitly?

**Comment by jakebailey** [comment]
I mean, I would say yes; I can invent any random context I want and then it may crash, but at least a cast indicates that I'm purposefully being unsafe. The latter case is of course "nice" because it looks cleaner, but is secretly hiding the fact that _nothing_ is restricting the context.

Maybe that's okay given mocking libraries already do this sort of thing, it just seems error-prone. Comparatively, `ava` recommends casting, `vitest` has an explicit context extension system (https://vitest.dev/guide/test-context#typescript). (I honestly don't know how this works in practice in Mocha; I wasn't able to quickly find a doc)

**Comment by joeyparrish** [comment]
> I mean, I would say yes; I can invent any random context I want and then it may crash, but at least a cast indicates that I'm purposefully being unsafe. The latter case is of course "nice" because it looks cleaner, but is secretly hiding the fact that nothing is restricting the context.

My way is more compact, but no more or less safe than a cast, and no more or less explicit.

> I honestly don't know how this works in practice in Mocha; I wasn't able to quickly find a doc

What Mocha does under the hood is ... weird.  I don't love it.  They define `Context`, then each `Suite` gets both its own Context instance and its own secret `Context` subclass for some reason... neither of which I understand.  The root Suite has an instance of `Context`, then a child Suite beneath that has an instance of a local `Context` that inherits from the parent's `Context` instance... it's really odd.

What I'm doing to extend the Context for my test suite is adding properties to it in a global `beforeAll`.  This is how/where you would do this without a subclass interface or strong types for your custom fixtures.

I just want to have those properties be strongly-typed when I refer to them, so I have something like:

```ts
interface TestContext extends Mocha.Context {
  doSomething(): void;
  aThing: number;
  // etc.
}

async function beforeAll(this: TestContext): Promise<void> {
  this.doSomething = function() {
    // ...
  };

  this.aThing = 5;
}

export const mochaHooks = {                                                      
  beforeAll,
};
```

It's somewhat hacky, yes, but it's a hack that is limited in scope to this one bootstrap file, and the actual tests get to trust in the extensions and refer to them with the correct types.  Mocha also doesn't offer anything better than this, frankly.  `vitest` has clearly thought this through.

The other thing is that Mocha, from plain JavaScript, doesn't actually have any type safety, and you're expected to extend Context dynamically if you need to.  The index signature here reflects that fact.  I'm just trying to wrangle that into extensions that can be easily type-checked.

**Comment by joeyparrish** [comment]
It seems there's no interest in this, and frankly, I can find other solutions.  (For example, move fixtures to a singleton and don't bother putting them on `Context` or subclassing it.)  I'll just let this one go.  Thanks, @jakebailey, for your time.

**Pull request closed without merging**
